### PR TITLE
add link_checker

### LINKED ISSUE #719: Test links in docs
Investigate a way to test hyperlinks in docs / markdown / notebooks to make sure they work

### PR DESCRIPTION
**Pull request by priyanshuone6**
# Description

Added workflow to test urls in the docs
Referred [this](https://github.com/marketplace/actions/urlchecker-action)

## Testing
Tested this workflow locally using [nektos/act](https://github.com/nektos/act)

snapshot

Fixes #719 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/master/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [x] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ flake8`
- [ ] All tests pass: `$ python run-tests.py --unit`
- [ ] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Review by tinosulzer — COMMENTED** [review]
Thanks @priyanshuone6 ! Did you run this locally (adding a known broken link) to check whether it does pick up broken links? I am slightly surprised that there aren't any broken links, though of course it's possible that we got lucky

File: .github/workflows/link_checker.yml
Comment (tinosulzer) [reply .github/workflows/link_checker.yml:20]: Nevermind, I read the link you shared now. Can you leave quiet mode off so that we can check each file has been checked?

Comment (priyanshuone6) [reply .github/workflows/link_checker.yml:20]: You are right, they both are quite contradictory, I'll remove the quiet mode.

**Comment by priyanshuone6** [comment]
I tried to test this workflow locally by using [nketos/act](https://github.com/nektos/act) and found that when I try to pass more than one extension in `file-extension`, in our case it's `.rst` and `.md`, it stops checking links and instead passes them as all correct.

**Comment by priyanshuone6** [comment]
I tried using https://github.com/urlstechie/urlchecker-action and found it better, it works with different file formats like `.py`, `.rst`, `.md` and it differentiates valid and invalid links using green and red.

**Comment by tinosulzer** [comment]
Ok great. Would be good to also test the notebooks (`.ipynb`)

**Comment by tlestang** [comment]
Thanks @priyanshuone6 , I tried that workflow on my fork and I noticed the output from url_checker doesn't display colors, see https://github.com/tlestang/PyBaMM/runs/1782502527?check_suite_focus=true
This makes it impossible to separate valid links from broken links. 

Another thing is that urls embedded links in SVG inside notebooks are picked up, but with a trailing backslash. 
Example
```
/github/workspace/examples/notebooks/solution-data-and-processed-variables.ipynb 
 --------------------------------------------------------------------------------
https://matplotlib.org/
http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\
http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\
http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\
http://www.w3.org/1999/xlink\
http://www.w3.org/1999/02/22-rdf-syntax-ns#\
```
Could you maybe specify the `exclude_patterns` option to exclude such links for the search?

**Comment by priyanshuone6** [comment]
@tlestang I searched about it and found that github actions do not currently support colors, but in the end, we do get all the invalid links together, look at [line 1126](https://github.com/tlestang/PyBaMM/runs/1782502527?check_suite_focus=true#step:4:1126)
I will look into the embedded links in SVG and try to fix them using exclude pattern.

**Comment by tinosulzer** [comment]
This seems to be working now with a manageable number of broken links reported. Weirdly, some of the broken links seem to work fine, so I don't know why URL-checker is reporting they are broken. Let me know if you need help finding working links to replace the broken ones with

**Comment by tinosulzer** [comment]
You can ignore the failing benchmarks test. @tlestang can we exclude that test when it comes from a fork?

**Comment by priyanshuone6** [comment]
Some urls starting with https://github.com/pybamm-team/PyBaMM/pull/ in `CHANGELOG.md` are being marked as invalid, I tested with different timeout settings to mark as many links valid as it can, and found 10 seconds to be perfect in terms of time and number of links.
Should I include https://github.com/pybamm-team/PyBaMM/pull/ in `exclude_patterns` or maybe exclude the whole `CHANGELOG.md`?
Also, should I remove the option to run this workflow in pull requests because every time someone will make a PR, and if it fails due to these unavoidable reasons, a red cross will appear in the PR due to that and it may confuse others?

**Comment by tinosulzer** [comment]
Yes, perhaps the way to go is to only run the url check on the cron job. The odds of someone adding a broken link are very low, what usually happens is a link changes and we don't update it because (until now) there was no automatic way of checking all the links. That way, you can also increase the timeout and it doesn't matter if the cron job takes a while. I wouldn't be totally against excluding `CHANGELOG.md` either since it is usually pretty easy to see if those links are broken.

However, there are some links [here](https://github.com/pybamm-team/PyBaMM/pull/1347/checks?check_run_id=1784001718#step:4:1059) that are actually broken and need to be fixed before this branch can be merged. Most of them are from changes to pybamm branch names or file structure and it should be fairly easy to find where the file is now (e.g. control+p in VSCode)

**Comment by priyanshuone6** [comment]
@tinosulzer Could you please help me with these broken links
https://pybamm.readthedocs.io/en/latest/source/models/lead_acid/composite.html
https://github.com/pybamm-team/PyBaMM/blob/master/examples/notebooks/add-model.ipynb

**Comment by tinosulzer** [comment]
First one should point here: https://pybamm.readthedocs.io/en/latest/source/models/lead_acid/higher_order.html
Second one should point here: https://github.com/pybamm-team/PyBaMM/tree/develop/examples/notebooks/Creating%20Models

**Review by tinosulzer — APPROVED** [review]
Thanks! If all the broken links are fixed (can't see the test now) then I am happy to merge once the other tests pass

**Comment by priyanshuone6** [comment]
I have changed the `timeout` back to 10 seconds, I observed that making it 20 seconds adds more valid links in the "not pass" category.
Please check [this](https://github.com/priyanshuone6/PyBaMM/runs/1801728747?check_suite_focus=true#step:4:1046) for the output

**Comment by tinosulzer** [comment]
@all-contributors add @priyanshuone6  for tests

**Pull request merged**
